src/c1_koiter_steigmann/koiter_steigmann_curvable_bell_elements.h

```diff
@@ -9,580 +9,1183 @@
-        jac_inv(0, 1) = -ti[0] / det;
-        jac_inv(1, 0) = -ni[1] / det;
-        jac_inv(1, 1) = ni[0] / det;
-
-        // Fill out the Hessian
-        // (unneeded -- can calculate the inverse components by hand)
-        for (unsigned alpha = 0; alpha < 2; alpha++)
-        {
-          // hess[alpha](0,0) = 0.0;
-          hess[alpha](0, 1) = dnids[alpha];
-          hess[alpha](1, 0) = dnids[alpha];
-          hess[alpha](1, 1) = dtids[alpha];
-        }
+          // Get whether each value is pinned
+          bool right_ui_pinned = right_data_pt->is_pinned(right_ui_index);
+          bool left_ui_pinned = left_data_pt->is_pinned(left_ui_index);
 
-        // Fill out inverse of Hessian
-        // H^{-1}abg = J^{-1}ad Hdez J^{-1}eb J^{-1}zg
-        for (unsigned alpha = 0; alpha < 2; alpha++)
-        {
-          for (unsigned beta = 0; beta < 2; beta++)
+          // If anything is free, mark it as used and continue without doing
+          // anything else
+          if (!right_ui_pinned && !right_data_used[right_ui_index])
           {
-            for (unsigned gamma = 0; gamma < 2; gamma++)
-            {
-              for (unsigned alpha2 = 0; alpha2 < 2; alpha2++)
-              {
-                for (unsigned beta2 = 0; beta2 < 2; beta2++)
-                {
-                  for (unsigned gamma2 = 0; gamma2 < 2; gamma2++)
-                  {
-                    hess_inv[alpha](beta, gamma) -=
-                      jac_inv(alpha, alpha2) * hess[alpha2](beta2, gamma2) *
-                      jac_inv(beta2, beta) * jac_inv(gamma2, gamma);
-                  }
-                }
-              }
-            }
+            // [zdec] debug
+            std::cout << "eqn " << condition_index
+		      << " depends on R dof" << right_ui_index
+		      << std::endl;
+            right_data_used[right_ui_index] = true;
           }
-        }
+          else if (!left_ui_pinned && !left_data_used[left_ui_index])
+          {
+            // [zdec] debug
+            std::cout << "eqn " << condition_index
+		      << " depends on L dof" << left_ui_index
+		      << std::endl;
+            left_data_used[left_ui_index] = true;
+          }
+          else
+          {
+            // --------------------------------------------------------------
 
-        // Fill in the rotation matrix using the new basis
-        fill_in_rotation_matrix_at_node_with_basis(j_node, jac_inv, hess_inv);
+            // [zdec] Below has to be true before solving here as there is no
+            // possibility that free dofs are being used elsewhere
 
-      } // end loop over vertices
-    } // end of update_rotation_matrices()
+            // If we made it here, it is because all dofs in the constraint are
+            // pinned so we need to check the constraint is satisfied manually
+            // and then remove it by pinning the corresponding lagrange
+            // multiplier
 
+            // hierher Aidan: should this remain alive?
```

Comment:
Answered earlier.

Reply:
Aidan: please tidy up brutally; I'll keep this comment in there to remind you but will mark this as resolved